### PR TITLE
gitserver: add functionality to run git commands locally

### PR DESCRIPTION
**Pull request by sashaostrikov**
This PR is another step towards making it possible to run tests without gitserver server instance running, which is needed for further refactoring (https://github.com/sourcegraph/sourcegraph/issues/33280)

This allows using GitCommand interface implementation (LocalGitCommand) for tests which don't need a running Gitserver instance.

Closes https://github.com/sourcegraph/sourcegraph/issues/34258

## Test plan
New test case added

### CODE REVIEW

**Review by mollylogue — APPROVED** [review]

File: internal/gitserver/client.go [hunk 1]
```diff
@@ -954,6 +957,96 @@ func (c *Cmd) StdoutReader(ctx context.Context) (io.ReadCloser, error) {
 	}, nil
 }
 
+// LocalGitCommand is a GitCommand interface implementation which runs git commands against local file system.
+//
+// This struct uses composition with exec.Cmd which already provides all necessary means to run commands against
+// local system.
+type LocalGitCommand struct {
```
Comment (vrto) [inline internal/gitserver/client.go:964]: Will there be a non-local flavor of GitCommand as a follow-up PR?

**Review by vrto — APPROVED** [review]
Awesome to see this happening! Left one small question :)

File: internal/gitserver/client_test.go [hunk 2]
```diff
@@ -735,3 +735,47 @@ func TestClient_BatchLog(t *testing.T) {
...
 	}
 }
+
+func TestLocalGitCommand(t *testing.T) {
+	// creating a repo with 1 committed file
+	root, err := os.MkdirTemp("", t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(root)
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/gitserver/client_test.go:745]: That's a great proposal!
I cannot really think of a convenient way to do it from the back of my head, but I can see that such pattern can be seen [in many places](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+defer+os.RemoveAll%28&patternType=literal) of our code.

I will try to come up with anything in another commit, which will add more tests!

File: internal/gitserver/client.go (hunk 1 above)
Comment (vrto) [reply internal/gitserver/client.go:964]: Should we then give it a more descriptive name? Like `RemoteGitCommand`? Or maybe let's go full Java on this and call it `GitCommandLocalAdapter`? :D

**Comment by camdencheek** [comment]
Oooh nice! I've been wanting this for a while

File: internal/gitserver/client_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanslade) [reply internal/gitserver/client_test.go:745]: We should be able to do it with a helper function that uses `t.Cleanup()` internally.

File: internal/gitserver/client.go (hunk 1 above)
Comment (ryanslade) [reply internal/gitserver/client.go:964]: Agree, I think `RemoveGitCommand` would be best since that's consistent with `LocalGitCommand`

**Review by ryanslade — COMMENTED** [review]
LGTM

File: internal/gitserver/client.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/gitserver/client.go:964]: > Or maybe let's go full Java on this and call it `GitCommandLocalAdapter`? :D 

I wanted to go full Java and use generics 😂 we are still waiting for them, although they are on its way in Sourcegraph

Comment (sashaostrikov) [reply internal/gitserver/client.go:964]: Next PR will fix the naming

**Pull request merged**